src/ldapclient_lib.py

```diff
@@ -0,0 +1,395 @@
+        self.api_port = snapshot["api_port"]
+        self.ldap_port = snapshot["ldap_port"]
+
+
+class ServerUnavailableEvent(EventBase):
+    """Charm Event triggered When the LDAP server is unavailable."""
+
+    def __init__(
+        self,
+        handle: Handle,
+    ):
+        super().__init__(handle)
+
+
+class LdapReadyEvent(EventBase):
+    """Charm Event triggered when LDAP is ready to start."""
+
+    def __init__(
+        self,
+        handle: Handle,
+    ):
+        super().__init__(handle)
+
+
+class LdapClientProviderCharmEvents(CharmEvents):
+    """Events the LDAP Client requirer can leverage."""
+
+    config_data_unavailable = EventSource(ConfigDataUnavailableEvent)
+    glauth_snap_ready = EventSource(GlauthSnapReadyEvent)
+    server_unavailable = EventSource(ServerUnavailableEvent)
+
+
+class LdapClientRequirerCharmEvents(CharmEvents):
+    """Events the LDAP Client requirer can leverage."""
+
+    certificate_available = EventSource(CertificateAvailableEvent)
+    certificate_unavailable = EventSource(CertificateUnavailableEvent)
+    config_data_available = EventSource(ConfigDataAvailableEvent)
+    config_data_unavailable = EventSource(ConfigDataUnavailableEvent)
+    server_unavailable = EventSource(ServerUnavailableEvent)
+    ldap_ready = EventSource(LdapReadyEvent)
+
+
+class LdapClientProvides(Object):
+    """Provides-side of the ldapclient integration."""
+
+    on = LdapClientProviderCharmEvents()
+
+    def __init__(self, charm: CharmBase, integration_name: str) -> None:
+        super().__init__(charm, integration_name)
+        self.framework.observe(
+            charm.on[integration_name].relation_broken,
+            self._on_relation_broken,
+        )
+        self.framework.observe(
+            charm.on[integration_name].relation_joined,
+            self._on_relation_joined,
+        )
+        self.charm = charm
+        self.integration_name = integration_name
+
+    def _get_hostname(self) -> str:
+        """Get GLAuth hostname.
+
+        Returns:
+            GLAuth hostname.
+        """
+        hostname = subprocess.run(
+            ["cat", "/etc/hostname"], capture_output=True, text=True
+        ).stdout.strip()
+        return hostname
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle relation-broken event.
+
+        When the ldapclient relation is broken and emits:
+        - Server unavailable event: When the ldap server can't be reached.
+        """
+        # Remove Obsolete Secrets
+        ca_cert_secret = self.model.get_secret(label="ca-cert")
+        ldbd_secret = self.model.get_secret(label="ldap-default-bind-dn")
+        lp_secret = self.model.get_secret(label="ldap-password")
+        ca_cert_secret.remove_all_revisions()
+        ldbd_secret.remove_all_revisions()
+        lp_secret.remove_all_revisions()
+        self.on.server_unavailable.emit()
+
+    def _on_relation_joined(self, event: RelationJoinedEvent) -> None:
```

Comment:
This function is DENSE which makes it slightly hard to read. Maybe adding a newline before each comment will make it easier to read.